### ISSUE TITLE
terragrunt plan-all - Add option to apply dependencies

### ISSUE THREAD

**Issue by markhopwoodnvm** [gruntwork-io/terragrunt#262]
We're implementing a pipeline around our terraform builds where each module has a CI build which runs a plan and apply. We'd like to be able to use this model on modules that have dependencies on other modules. This would require all dependencies to be applied prior to running a plan on the module we are testing. Using the example from your documentation :

root
├── backend-app
│   ├── main.tf
│   └── terraform.tfvars
├── frontend-app
│   ├── main.tf
│   └── terraform.tfvars
├── mysql
│   ├── main.tf
│   └── terraform.tfvars
├── redis
│   ├── main.tf
│   └── terraform.tfvars
└── vpc
    ├── main.tf
    └── terraform.tfvars

with the following dependencies between Terraform modules:

backend-app depends on mysql, redis, and vpc
frontend-app depends on backend-app and vpc
mysql depends on vpc
redis depends on vpc
vpc has no dependencies

terragrunt = {
  dependencies {
    paths = ["../vpc", "../mysql", "../redis"]
  }
}

A terragrunt plan-all (with apply dependencies opiton) for backend-app would run the apply on vpc, mysql and redis, then a plan on backend-app

**Comment by brikis98** [comment]
I think the expectation for the `plan` command is that it's read-only and therefore always safe to run. It may be a bit surprising for it to be able to apply changes, even if you ask for them explicitly via a flag.

Perhaps the proper abstraction for this is an `apply-all-dependencies` command that runs `apply-all` on the dependencies of the current module, and their dependencies recursively, but not the current module?

**Comment by markhopwoodnvm** [comment]
Yep, that sounds good

**Comment by vas1468** [comment]
Can we give some love to this enhancement please? No one is assigned yet :( Unfortunately I cannot help.

**Comment by w32-blaster** [comment]
+1

**Comment by Hmerac** [comment]
I also encountered with this, some of our modules have dependencies over other modules.

The problem is, at first-run of **terragrunt plan-all** phase, dependent modules aren't able to retrieve output variables from other module's remote state file because **terragrunt plan-all** doesn't initialize state files as you know.

We needed to specify **terraform apply** for each module with regard to dependencies. Even this doesn't solve the fail problem when you run **terragrunt plan-all** at first-run.

So -> +1 🙌

**Comment by brikis98** [comment]
It says it right in the [README](https://github.com/gruntwork-io/terragrunt#the-apply-all-destroy-all-output-all-and-plan-all-commands):

> Note: It is important to realize that you could get errors running plan-all if you have dependencies between your projects and some of those dependencies haven't been applied yet.
>
> Ex: If module A depends on module B and module B hasn't been applied yet, then plan-all will show the plan for B, but exit with an error when trying to show the plan for A.

**Comment by brikis98** [comment]
Hm, that's weird. I was responding to a comment that's now gone... Uh... I guess the commenter deleted it?

**Comment by Xtigyro** [comment]
Hey folks - has there been any progress since the last comment?

**Comment by brikis98** [comment]
Not that I'm aware of. PRs still welcome.

**Comment by derekrprice** [comment]
Not sure how complicated this would be, but it seems like the right way to approach this would be not to actually apply the dependencies, but to fake it.  As each level is planned, the desired state is already output.  That desired state could be used to create a mock remote state used by each dependent plan.  Terragrunt is already manipulating remote state in order to share a single definition block down the tree.  Is there a way that it could substitute a "terragrunt-mock" provider in the specific case of plan-all?

In my defense, if this is a crazy idea, I don't have any direct experience with the Terragrunt code base or Go yet.  Maybe soon, if I keep finding issues like this.  :)

**Comment by brikis98** [comment]
@derekrprice Don't think that would work. Figuring out the right fake outputs to return for every modules is hard enough, but if we're just returning mock data, then when those outputs are fed as inputs to other modules so those modules can run `plan`, they'll hit errors because those outputs won't correspond to real resources, so all the API calls will fail.

**Comment by derekrprice** [comment]
@brikis98 Vanilla Terraform can do it when all the resources are in the same state file. It seems like it should be possible. Not easy, perhaps, but possible.

**Comment by trallnag** [comment]
@derekrprice I feel like this would be out of scope for Terragrunt. It basically means reinventing Terraform

**Comment by marshall7m** [comment]
Here’s a Terraform module that detects all changes within Terragrunt dependency chains and orchestrates a continuous deployment flow that respects the dependeny chain order. The module offers AWS account-level dependencies and approval requirements. In addition, it includes handling for rolling back newly deployed Terraform provider resources. See the README.md for the full description of how this all works [here](https://github.com/marshall7m/terraform-aws-infrastructure-live-ci).

Side note: I’m the owner of this repository. If you do check it out, I'm open to any questions and brutally honest feedback!

**Comment by ikegentz** [comment]
It's not a perfect solution (IMO a "perfect" solution is way beyond the scope of Terragrunt), but the `mock_outputs` feature of Terragrunt is about as close as you can get to solving this without re-inventing the wheel.

**Comment by fraz78** [comment]
> It's not a perfect solution (IMO a "perfect" solution is way beyond the scope of Terragrunt), but the `mock_outputs` feature of Terragrunt is about as close as you can get to solving this without re-inventing the wheel, but this still doesn't solve the entire issue. Sometimes there are modules that have a `data` resource in them which will **always** run and need actual outputs (not just mock outputs), so the use-case that OP is suggesting is still valid. Otherwise as @dmead mentioned, you have to write some bash-foo to achieve this (we're using makefiles)

I'm having this same issue with a data call when running a `terragrunt run-all plan`. Any advice?

**Comment by ikegentz** [comment]
@fraz78 Since the time I had made that comment we just accepted that `run-all plan` doesn't work if you haven't already done an initial `run-all apply`, and just used Terragrunt as such. That said, the only real suggestion I'd make is to create a makefile/script of some sort in each of your terragrunt directories that does an `apply` on any resources you have identified as causing the issue. Then you run `make plan-all` (which under the hood does those applies/plans as necessary. 

Not trying to dissuade you from attempting that implementation, but we found the juice wasn't worth the squeeze, and just didn't try to use terragrunt that way.

**Comment by tspearconquest** [comment]
Personally, I run into this issue only after I have created the resources and want to add a new one to an already existing module.

In my case, I have a module for creating managed identity in azure, and a separate module for creating a managed identity federated credential. The latter depends on the former.

Before creating resources, I leverage mock dependencies, and my plans all look fine. I then create the resources with run-all apply. Then later I needed to add a new MID and federated credential, so I add the resource to the first module, and I add a mock dependency to the second module, along with the resource in the second module.

The plan for the MID looks good but the plan for the federated credential fails because the dependency doesn't exist.

What I expected to happen is that terragrunt would merge my new mock dependency definition into the existing real resource dependency during the plan, so that the plan for the federated credential properly functions as if the real MID was already created. Instead, terragrunt ignores the mock dependency entirely (seemingly because the other MIDs are already present.)

If I just do run-all apply, it works fine, and if I destroy everything and then run a plan for the updated code, it also works fine.

So, to me, terragrunt should figure out which dependencies really exist already and which don't, and merge in any mock dependencies that don't exist but are defined as mock dependencies to help generate a proper plan.

**Comment by MattCosturos** [comment]
> What I expected to happen is that terragrunt would merge my new mock dependency definition into the existing real resource dependency during the plan, so that the plan for the federated credential properly functions as if the real MID was already created. Instead, terragrunt ignores the mock dependency entirely (seemingly because the other MIDs are already present.)

Are you using `mock_outputs_merge_with_state = true` in your dependency definition? The default is to NOT merge the new mock outputs with the existing state. We run into this a lot.

**Comment by Speeddymon** [comment]
> Are you using `mock_outputs_merge_with_state = true` in your dependency definition?

I wasn't aware of this. Will give it a try. Thanks!
